### PR TITLE
variables playerName and playerColor are now writable

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
We discussed this in a meeting long, long ago.

Please note that this can only set the name and color of the player triggering the routine.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3560/pr-test (or any other room on that server)

**Comment by ArnoldSmith86** [comment]
Aaaand someone already managed to set `playerName` to `null`. :rofl:

**Comment by bjalder26** [comment]
> Aaaand someone already managed to set `playerName` to `null`. 🤣

Yeah, I wanted to make a random name generator, but I'm not done (it doesn't work yet), and now I have to run for a bit.

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -1049,6 +1049,12 @@ export class Widget extends StateManaged {
       showOverlay('debugButtonOverlay');
 
     batchEnd();
+
+    if(variables.playerColor != playerColor && typeof variables.playerColor == 'string')
```
Comment (robartsd) [inline client/js/widgets/widget.js:1053]: Please check that the string matches  `/^#[0-9a-fA-F]{6}$/`.

**Review by robartsd — CHANGES_REQUESTED** [review]

**Comment by bjalder26** [comment]
I made a random name & color generator and noticed something odd.  Only one of them would update at a time.  It turned out that I was able to resolve that by putting the lines that set those variables together instead of having them separated by a bunch of computations. 

To be more specific, calculating a random color required a lot of computations, and the rando color would only update if the name did NOT update (was set to the same name).  Then when I moved the lines together to set those variables it worked.  Besides computations, there were a bunch of IF statements between them, so that might have been the issue too.

**Comment by ArnoldSmith86** [comment]
Always keep a copy of the buggy ones, please.

**Comment by bjalder26** [comment]
> Always keep a copy of the buggy ones, please.

I put it back in the room and limited it to two names so the chance of it 'bugging' is 50/50.

**Comment by bjalder26** [comment]
I did some more testing, and it is the IF separating the two variable changes that somehow prevents the two variables from changing at the same time.  Further, if the playerColor is the one before the `IF` statement, then it will never change, only the playerName will change.

I am guessing that the playerColor 'triggers' with the wrong playerName (or possibly no playerName) if it is separated from the playerName by an `IF` statement in the JSON.

**Comment by robartsd** [comment]
> I did some more testing, and it is the IF separating the two variable changes that somehow prevents the two variables from changing at the same time. Further, if the playerColor is the one before the `IF` statement, then it will never change, only the playerName will change.
> 
> I am guessing that the playerColor 'triggers' with the wrong playerName (or possibly no playerName) if it is separated from the playerName by an `IF` statement in the JSON.
> 
> Of course, you can just not put an `IF` statement between them.

Sounds like the variables are getting written to state at the end of any (sub)routine not just the outermost routine. This causes issues when returning to the calling routine if the calling routine later interacts with the variables.

**Comment by robartsd** [comment]
When we call the `thenRoutine` or `elseRoutine` all variables including the changed playerColor and playerName variables get passed to the routine. When the branch completes, the changes are queued up to be sent to the server rather than waiting to be queued up in the parent routine.

**Comment by ArnoldSmith86** [comment]
Yeah, I'm pretty sure I know how to fix it (locally also setting `playerName`). I just didn't have time yet.

**Comment by ArnoldSmith86** [comment]
Please check if the issues have been resolved.

**Review by bjalder26 — APPROVED** [review]
My intentionally buggy name generators now work fine.  I can't find any issues.  Approved based on testing.

**Review by robartsd — APPROVED** [review]

**Comment by ArnoldSmith86** [comment]
I was about to merge but I noticed a crash when player cursors are still updated with the old name or something like that. Gotta check..

**Comment by bjalder26** [comment]
> I was about to merge but I noticed a crash when player cursors are still updated with the old name or something like that. Gotta check..

Could it be when the player has gone inactive, then goes back and clicks a button to choose a new name?  I thought that might be the case, but I also assumed that was 'normal', since the page was probably 'stale' (with possible changes) and needed to be refreshed anyway.

**Comment by RaphaelAlvez** [comment]
Can you add the tag to? It will need to be writable anyway so maybe that's a good opportunity to add it. And maybe also rename the color to remove "team" from there.
I got really confused when I was looking at that code

**Comment by ArnoldSmith86** [comment]
> I also assumed that was 'normal', since the page was probably 'stale' (with possible changes) and needed to be refreshed anyway.

If the complete page reloads, that means that the client had an unexpected error. It only reloads so that the players aren't left with a broken page. So ideally that should never happen.

> Can you add the tag to?

No, not now.

> And maybe also rename the color to remove "team" from there.
> I got really confused when I was looking at that code

Initially my idea was that the color is the team. So two red players would be the same team. I don't think that's a good idea anymore though.

**Comment by bjalder26** [comment]
> complete page reloads, that means that the client had an unexpected error. It only reloads so that the players aren't left with a broken page. So ideally that should never happen.

I think I may have found what causes it.  I opened the test room on my phone.  I can't see anything on my phone except what appears to be a blank screen.  If I scroll on my phone the page reloads on my laptop.

http://212.47.248.129:3560/pr-test

**Comment by robartsd** [comment]
> Initially my idea was that the color is the team. So two red players would be the same team. I don't think that's a good idea anymore though.

I'd like to copy the idea of player number or initials in the cursor circle from PICO. If we did that then the cursor could represent the team with color and distinguish team members with the initials.

**Comment by 96LawDawg** [comment]
I added a comment about how to set `playerName` and `playerColor` in the Functions section of the wiki.

**Pull request merged**